File: android/src/main/java/com/paxposlink2/PaxPoslink2Module.java
package com.paxposlink2;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;
import java.util.Map;
import java.util.HashMap;
import com.pax.poscore.commsetting.AidlSetting;
// import com.pax.poslink.POSLinkAndroid; //TODO: NEED TO FIX IMPORT FOR POSLinkAndroid
import com.pax.poscore.LogSetting;
// import com.pax.poslink.PosLink;  //TODO: NEED TO FIX IMPORT FOR POSLinkAndroid
// import com.pax.poslink.PaymentRequest;  //TODO: NEED TO FIX IMPORT FOR POSLinkAndroid
// import com.pax.poslink.ProcessTransResult;  //TODO: NEED TO FIX IMPORT FOR POSLinkAndroid
// import com.pax.poslink.PaymentResponse;  //TODO: NEED TO FIX IMPORT FOR POSLinkAndroid
// import com.pax.poslink.BatchRequest;  //TODO: NEED TO FIX IMPORT FOR POSLinkAndroid
// import com.pax.poslink.BatchResponse;  //TODO: NEED TO FIX IMPORT FOR POSLinkAndroid
// import com.pax.poslink.CommSetting;  //TODO: NEED TO FIX IMPORT FOR POSLinkAndroid
import android.os.HandlerThread;
import android.os.Handler;
import android.os.Looper;
import java.lang.Runnable;
import android.util.Log;
import android.content.Context;

@ReactModule(name = PaxPoslink2Module.NAME)
public class PaxPoslink2Module extends ReactContextBaseJavaModule {
  public static final String NAME = "PaxPoslink2";
//  private PosLink poslink;
  private HandlerThread handlerThread;
  private Context context;

  public PaxPoslink2Module(ReactApplicationContext reactContext) {

    super(reactContext);
    this.context = reactContext;

  }

  @Override
  @NonNull
  public String getName() {
    return NAME;
  }


  @ReactMethod
  public void initPOSLink(String type, String timeout, Boolean proxy, String ip, String port, Promise promise) {

    try {
      LogSetting logSetting = new LogSetting();

      logSetting.setLogMode(true); // Open or close log
      logSetting.setLevel(LogSetting.LOGLEVEL.DEBUG); // Set log level
      logSetting.setLogFileName("POSLinkLog"); // Set the file name of output log
      logSetting.setOutputPath("sdcard/log"); // Set path for output log
      logSetting.setLogDays("30"); // Keep log for 30 days

      // Setting connection configurations
      CommunicationSetting commSetting = new AidlSetting();

      commSetting.setType(type);

      if (ip != null) {
        commSetting.setDestIP(ip);
        commSetting.setDestPort(port);
      }

      commSetting.setTimeOut(timeout);

      if (proxy) {
        commSetting.setEnableProxy(proxy);
      }

      poslink = new PosLink(this.context);

      poslink.SetCommSetting(commSetting);

      promise.resolve(true);

    } catch (Exception e) {
      promise.reject("Exception Error", e);
    }

  }

  @ReactMethod
  private void runPayment(String amount, String tip, Promise promise) {

    PaymentRequest req = new PaymentRequest();

    req.TenderType = req.ParseTenderType("CREDIT");
    req.TransType = req.ParseTransType("SALE");
    req.ECRRefNum = "1";
    req.Amount = amount;
    req.TipAmt = tip;

    this.runTransaction(req, promise);

  }

  @ReactMethod
  private void runAuth(String amount, Promise promise) {

    PaymentRequest req = new PaymentRequest();

    req.TenderType = req.ParseTenderType("CREDIT");
    req.TransType = req.ParseTransType("AUTH");
    req.ECRRefNum = "1";
    req.Amount = amount;

    this.runTransaction(req, promise);

  }

  @ReactMethod
  private void runReturn(String amount, Promise promise) {

    PaymentRequest req = new PaymentRequest();

    req.TenderType = req.ParseTenderType("CREDIT");
    req.TransType = req.ParseTransType("RETURN");
    req.ECRRefNum = "5";
    req.Amount = amount;

    this.runTransaction(req, promise);

  }

  @ReactMethod
  private void closeBatch(Promise promise) {

    handlerThread = new HandlerThread("MyHandlerThread");
    handlerThread.start();
    Looper looper = handlerThread.getLooper();
    Handler handler = new Handler(looper);

    handler.post(new Runnable() {

      @Override
      public void run() {
        BatchRequest batchRequest = new BatchRequest();
        batchRequest.TransType = batchRequest.ParseTransType("BATCHCLOSE");
        poslink.BatchRequest = batchRequest;

        ProcessTransResult ptr = poslink.ProcessTrans();

        if (ptr.Code == ProcessTransResult.ProcessTransResultCode.OK) {
          BatchResponse response = poslink.BatchResponse;

          WritableMap map = Arguments.createMap();
          map.putString("AuthCode", response.AuthCode);
          map.putString("BatchFailedCount", response.BatchFailedCount);
          map.putString("BatchFailedRefNum", response.BatchFailedRefNum);
          map.putString("BatchNum", response.BatchNum);
          map.putString("CashAmount", response.CashAmount);
          map.putString("CashCount", response.CashCount);
          map.putString("CHECKAmount", response.CHECKAmount);
          map.putString("CreditAmount", response.CreditAmount);
          map.putString("CreditCount", response.CreditCount);
          map.putString("DebitAmount", response.DebitAmount);
          map.putString("DebitCount", response.DebitCount);
          map.putString("EBTAmount", response.EBTAmount);
          map.putString("EBTCount", response.EBTCount);
          map.putString("ExtData", response.ExtData);
          map.putString("GiftAmount", response.GiftAmount);
          map.putString("GiftCount", response.GiftCount);
          map.putString("HostCode", response.HostCode);
          map.putString("HostResponse", response.HostResponse);
          map.putString("HostTraceNum", response.HostTraceNum);
          map.putString("LoyaltyAmount", response.LoyaltyAmount);
          map.putString("LoyaltyCount", response.LoyaltyCount);
          map.putString("Message", response.Message);
          map.putString("ResultCode", response.ResultCode);
          map.putString("ResultTxt", response.ResultTxt);
          map.putString("SAFDeletedCount", response.SAFDeletedCount);
          map.putString("SAFFailedCount", response.SAFFailedCount);
          map.putString("SAFFailedTotal", response.SAFFailedTotal);
          map.putString("SAFTotalAmount", response.SAFTotalAmount);
          map.putString("SAFTotalCount", response.SAFTotalCount);
          map.putString("SAFUploadedAmount", response.SAFUploadedAmount);
          map.putString("SAFUploadedCount", response.SAFUploadedCount);
          map.putString("TID", response.TID);
          map.putString("Timestamp", response.Timestamp);

          promise.resolve(map);

        } else if (ptr.Code == ProcessTransResult.ProcessTransResultCode.TimeOut) {
          String errorMsg = ptr.Msg;
          String errorCode = String.valueOf(ptr.Code);
          promise.reject(ptr.Msg);

        } else {
          String errorMsg = ptr.Msg;
          String errorCode = String.valueOf(ptr.Code);
          promise.reject(ptr.Msg);
        }
      }
    });
  }

  @ReactMethod
  private void runPostAuth(String amount, String refNum, String authCode, Promise promise) {
    PaymentRequest req = new PaymentRequest();

    req.TenderType = req.ParseTenderType("CREDIT");
    req.TransType = req.ParseTransType("POSTAUTH");
    req.ECRRefNum = "1";
    req.OrigRefNum = refNum;
    req.AuthCode = authCode;
    req.Amount = amount;

    this.runTransaction(req, promise);
  }

  @ReactMethod
  private void runVoid(String refNum, String authCode, Promise promise) {
    PaymentRequest req = new PaymentRequest();

    req.TenderType = req.ParseTenderType("CREDIT");
    req.TransType = req.ParseTransType("VOID");
    req.ECRRefNum = "1";
    req.OrigRefNum = refNum;
    req.AuthCode = authCode;

    this.runTransaction(req, promise);
  }

  @ReactMethod
  private void runAdjustment(String amount, String refNum, Promise promise) {
    PaymentRequest req = new PaymentRequest();

    req.TenderType = req.ParseTenderType("CREDIT");
    req.TransType = req.ParseTransType("ADJUST");
    req.OrigTraceNum = refNum;
    req.ECRRefNum = "1";
    req.OrigRefNum = refNum;
    req.Amount = amount;

    this.runTransaction(req, promise);
  }

  @ReactMethod
  private void runTransaction(PaymentRequest request, Promise promise) {

    handlerThread = new HandlerThread("MyHandlerThread");
    handlerThread.start();
    Looper looper = handlerThread.getLooper();
    Handler handler = new Handler(looper);

    handler.post(new Runnable() {

      @Override
      public void run() {
        Log.d("DEBUG_POSLINK_INFO", String.valueOf(poslink.GetCommSetting().getDestIP()));

        poslink.PaymentRequest = request;

        ProcessTransResult ptr = poslink.ProcessTrans();

        Log.d("DEBUG_PAYMENT_REQUEST", String.valueOf(ptr.Msg));

        // Check transaction result
        if (ptr.Code == ProcessTransResult.ProcessTransResultCode.OK) {
          /**
           * When the transResult.Code is OK, then the response has already been
           * assigned to poslink instance's PaymentResponse field automatically,
           * what you only need to do is get the response from the field
           */
          PaymentResponse response = poslink.PaymentResponse;

          Log.d("PAYMENT_RESPONSE_DEBUG AuthCode", response.AuthCode);
          WritableMap map = Arguments.createMap();
          map.putString("ApprovedAmount", response.ApprovedAmount);
          map.putString("ApprovedCashBackAmount", response.ApprovedCashBackAmount);
          map.putString("ApprovedMerchantFee", response.ApprovedMerchantFee);
          map.putString("ApprovedTaxAmount", response.ApprovedTaxAmount);
          map.putString("ApprovedTipAmount", response.ApprovedTipAmount);
          map.putString("AuthCode", response.AuthCode);
          map.putString("AuthorizationResponse", response.AuthorizationResponse);
          map.putString("AvsResponse", response.AvsResponse);
          map.putString("BogusAccountNum", response.BogusAccountNum);
          map.putString("CardType", response.CardType);
          map.putString("CvResponse", response.CvResponse);
          map.putString("DebitAccountType", response.DebitAccountType);
          map.putString("ECRTransID", response.ECRTransID);
          map.putString("EDCType", response.EDCType);
          map.putString("ExtraBalance", response.ExtraBalance);
          map.putString("GatewayTransactionID", response.GatewayTransactionID);
          map.putString("GiftCardType", response.GiftCardType);
          map.putString("HostAccount", response.HostAccount);
          map.putString("HostCardType", response.HostCardType);
          map.putString("HostCode", response.HostCode);
          map.putString("HostDetailedMessage", response.HostDetailedMessage);
          map.putString("HostResponse", response.HostResponse);
          map.putString("HostTimeStamp", response.HostTimeStamp);
          map.putString("IssuerResponseCode", response.IssuerResponseCode);
          map.putString("MaskedPAN", response.MaskedPAN);
          map.putString("Message", response.Message);
          map.putString("PayloadData", response.PayloadData);
          map.putString("PaymentAccountReferenceID", response.PaymentAccountReferenceID);
          map.putString("RawResponse", response.RawResponse);
          map.putString("RefNum", response.RefNum);
          map.putString("RemainingBalance", response.RemainingBalance);
          map.putString("RequestedAmount", response.RequestedAmount);
          map.putString("ResultCode", response.ResultCode);
          map.putString("ResultTxt", response.ResultTxt);
          map.putString("RetrievalReferenceNumber", response.RetrievalReferenceNumber);
          map.putString("SigFileName", response.SigFileName);
          map.putString("SignData", response.SignData);
          map.putString("Timestamp", response.Timestamp);
          map.putString("Track1Data", response.Track1Data);
          map.putString("Track2Data", response.Track2Data);
          map.putString("Track3Data", response.Track3Data);
          map.putString("TransactionIntegrityClass", response.TransactionIntegrityClass);
          map.putString("TransactionRemainingAmount", response.TransactionRemainingAmount);
          map.putString("ExtData", response.ExtData);

          promise.resolve(map);

        } else if (ptr.Code == ProcessTransResult.ProcessTransResultCode.TimeOut) {
          String errorMsg = ptr.Msg;
          String errorCode = String.valueOf(ptr.Code);
          Log.d("PAYMENT_RESPONSE_DEBUG", errorMsg);
          Log.d("PAYMENT_RESPONSE_DEBUG", errorCode);
          handler.getLooper().quit();
          promise.reject(ptr.Msg);

        } else {
          String errorMsg = ptr.Msg;
          String errorCode = String.valueOf(ptr.Code);
          Log.d("PAYMENT_RESPONSE_DEBUG", errorCode);
          Log.d("PAYMENT_RESPONSE_DEBUG", errorMsg);
          promise.reject(ptr.Msg);

          handler.getLooper().quit();

        }
      }
    });

  }

}

File: android/src/main/java/com/paxposlink2/PaxPoslink2Module.kt
package com.paxposlink2

import android.content.Context
import android.os.HandlerThread
import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.module.annotations.ReactModule
import com.pax.poscore.LogSetting
import com.pax.poscore.commsetting.AidlSetting
import com.pax.poscore.commsetting.BluetoothSetting
import com.pax.poscore.commsetting.CommunicationSetting
import com.pax.poscore.commsetting.TcpSetting
import com.pax.poscore.commsetting.UartSetting
import com.pax.poscore.commsetting.UsbSetting
import com.pax.poslinkadmin.ExecutionResult
import com.pax.poslinkadmin.constant.TransactionType
import com.pax.poslinkadmin.util.AmountRequest
import com.pax.poslinksemiintegration.POSLinkSemi
import com.pax.poslinksemiintegration.transaction.DoCashRequest
import com.pax.poslinksemiintegration.transaction.DoCashResponse
import com.pax.poslinksemiintegration.transaction.DoCreditRequest
import com.pax.poslinksemiintegration.transaction.DoCreditResponse
import com.pax.poslinksemiintegration.util.TraceRequest
import com.paxposlink2.PaxPoslink2Module

@ReactModule(name = PaxPoslink2Module.NAME)
class PaxPoslink2Module(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {
    private var poslink: POSLinkSemi? = null
    private val handlerThread: HandlerThread? = null
    private var communicationSetting: CommunicationSetting? = null
    private val context: Context = reactContext

    override fun getName(): String {
        return NAME
    }


    @ReactMethod
    fun initPOSLink(
        type: String,
        timeout: Int,
        nameOrMac: String?,
        ipOrSerial: String?,
        portOrBaud: String?,
        promise: Promise
    ) {
        try {
            val logSetting = LogSetting()
            logSetting.isEnable = true
            logSetting.level = LogSetting.LogLevel.DEBUG // Set log level
            logSetting.fileName = "POSLinkLog" // Set the file name of output log
            logSetting.filePath = "sdcard/log" // Set path for output log
            logSetting.setDays(30) // Keep log for 30 days

            poslink = POSLinkSemi.getInstance()
            poslink!!.setLogSetting(logSetting)
            initPaymentCommunication(type!!, timeout, nameOrMac, ipOrSerial, portOrBaud, promise)
        } catch (e: Exception) {
            promise.reject("Exception Error", e)
        }
    }

    @ReactMethod
    fun initPaymentCommunication(
        type: String,
        timeout: Int,
        nameOrMac: String?,
        ipOrSerial: String?,
        portOrBaud: String?,
        promise: Promise
    ) {
        // Setting connection configurations
        when (type) {
            "AIDL" -> {}
            "TCP" -> {
                val tcpSetting = TcpSetting(ipOrSerial, portOrBaud, timeout)
                this.communicationSetting = tcpSetting
            }

            "USB" -> {
                val usbSetting = UsbSetting()
                usbSetting.deviceName = nameOrMac
                this.communicationSetting = usbSetting
            }

            "BLUETOOTH" -> {
                val bluetoothSetting = BluetoothSetting()
                bluetoothSetting.macAddr = nameOrMac
                this.communicationSetting = bluetoothSetting
            }

            "AIDL" -> {
                val aidlSetting = AidlSetting()
                this.communicationSetting = aidlSetting
            }

            "UART" -> {
                val uartSetting = UartSetting(ipOrSerial, portOrBaud, timeout)
                this.communicationSetting = uartSetting
            }

            else -> {
                promise.reject("Invalid Type", type)
                return
            }
        }

        //      if (proxy) {
//        commSetting.setEnableProxy(proxy);
//      }
    }

    @ReactMethod
    fun makeCreditPayment(
        amount: String?,
        tip: String?,
        referenceNumber: String?,
        promise: Promise
    ) {
        val terminal = poslink!!.getTerminal(this.context, this.communicationSetting)

        val amountRequest = AmountRequest()
        amountRequest.transactionAmount = amount
        amountRequest.tipAmount = tip

        val traceRequest = TraceRequest()
        traceRequest.ecrReferenceNumber = referenceNumber

        val doCreditRequest = DoCreditRequest()
        doCreditRequest.transactionType = TransactionType.SALE
        doCreditRequest.amountInformation = amountRequest
        doCreditRequest.traceInformation = traceRequest

        var executionResult: ExecutionResult<DoCreditResponse>? = null

        executionResult = terminal.transaction.doCredit(doCreditRequest)

        this.handleCreditExecutionResult(executionResult, promise)
    }

    @ReactMethod
    fun makeCashPayment(amount: String?, tip: String?, referenceNumber: String?, promise: Promise) {
        val terminal = poslink!!.getTerminal(this.context, this.communicationSetting)

        val amountRequest = AmountRequest()
        amountRequest.transactionAmount = amount
        amountRequest.tipAmount = tip

        val traceRequest = TraceRequest()
        traceRequest.ecrReferenceNumber = referenceNumber

        val doCashRequest = DoCashRequest()
        doCashRequest.transactionType = TransactionType.SALE
        doCashRequest.amountInformation = amountRequest
        doCashRequest.traceInformation = traceRequest

        val executionResult: ExecutionResult<DoCashResponse> =
            terminal.transaction.doCash(doCashRequest)

        this.handleCashExecutionResult(executionResult, promise)
    }

//    fun makeReturnPayment(
//        amount: String?,
//        tip: String?,
//        referenceNumber: String?,
//        promise: Promise
//    ) {
//        val terminal = poslink!!.getTerminal(this.context, this.communicationSetting)
//
//        val amountRequest = AmountRequest()
//        amountRequest.transactionAmount = amount
//        amountRequest.tipAmount = tip
//
//        val traceRequest = TraceRequest()
//        traceRequest.ecrReferenceNumber = referenceNumber
//
//        val doCashRequest = DoCashRequest()
//        doCashRequest.transactionType = TransactionType.SALE
//        doCashRequest.amountInformation = amountRequest
//        doCashRequest.traceInformation = traceRequest
//
//        val executionResult: ExecutionResult<DoCashResponse> =
//            terminal.transaction.doCash(doCashRequest)
//
//        this.handleCashExecutionResult(executionResult, promise)
//    }

    private fun handleCreditExecutionResult(
        executionResult: ExecutionResult<DoCreditResponse>,
        promise: Promise
    ) {
        if (executionResult.isSuccessful) {
            val doCreditResponse = executionResult.response()
            Log.d("PAYMENT_RESPONSE_DEBUG AuthCode", doCreditResponse.responseCode())
            val map = Arguments.createMap()
            map.putString("ApprovedAmount", doCreditResponse.amountInformation().approvedAmount())
            map.putString(
                "ApprovedCashBackAmount",
                doCreditResponse.amountInformation().approvedCashBackAmount()
            )
            map.putString(
                "ApprovedMerchantFee",
                doCreditResponse.amountInformation().approvedMerchantFee()
            )
            map.putString(
                "ApprovedTaxAmount",
                doCreditResponse.amountInformation().approvedTaxAmount()
            )
            map.putString(
                "ApprovedTipAmount",
                doCreditResponse.amountInformation().approvedTipAmount()
            )
            map.putString("AuthCode", doCreditResponse.responseCode())
            map.putString("AuthorizationResponse", doCreditResponse.responseMessage())
            map.putString("AvsResponse", doCreditResponse.avsInformation().avsMessage())
            map.putString(
                "CurrentAccountNumber",
                doCreditResponse.accountInformation().currentAccountNumber()
            )
            map.putString("CardType", doCreditResponse.accountInformation().cardType().getName())
            map.putString(
                "CvdApprovalCode",
                doCreditResponse.accountInformation().cvdApprovalCode()
            )
            map.putString(
                "DebitAccountType",
                doCreditResponse.accountInformation().debitAccountType().getName()
            )
            map.putString("ECRTransID", doCreditResponse.hostTraceInformation().ecrTransactionId())
            map.putString("EDCType", doCreditResponse.edcType())
            //      map.putString("ExtraBalance", doCreditResponse.amountInformation().ExtraBalance);
            map.putString(
                "GatewayTransactionID",
                doCreditResponse.hostInformation().gatewayTransactionId()
            )
            map.putString("GiftCardType", doCreditResponse.accountInformation().giftCardType().getName())
            map.putString("ReferenceNumber", doCreditResponse.traceInformation().referenceNumber())
            map.putString("HostCardType", doCreditResponse.accountInformation().hostCardType())
            map.putString("HostResponseCode", doCreditResponse.hostInformation().hostResponseCode())
            map.putString(
                "HostResponseMessage",
                doCreditResponse.torInformation().hostResponseMessage()
            )
            map.putString(
                "HostDetailMessage",
                doCreditResponse.hostInformation().hostDetailedMessage()
            )
            map.putString("HostTimeStamp", doCreditResponse.hostTraceInformation().hostTimeStamp())
            map.putString(
                "HostReferenceNumber",
                doCreditResponse.hostTraceInformation().hostReferenceNumber()
            )
            map.putString(
                "IssuerResponseCode",
                doCreditResponse.hostInformation().issuerResponseCode()
            )
            map.putString("MaskedPAN", doCreditResponse.torInformation().maskedPan())
            map.putString("ResponseMessage", doCreditResponse.responseMessage())
            map.putString("PayloadData", doCreditResponse.payloadData())
            map.putString(
                "PaymentAccountReferenceID",
                doCreditResponse.hostInformation().paymentAccountReferenceId()
            )
            //      map.putString("RawResponse", doCreditResponse.RawResponse);
//      map.putString("RemainingBalance", doCreditResponse.RemainingBalance);
            map.putString("RequestedAmount", doCreditResponse.torInformation().originalAmount())
            //      map.putString("ResultCode", doCreditResponse.ResultCode);
//      map.putString("ResultTxt", doCreditResponse.ResultTxt);
            map.putString(
                "RetrievalReferenceNumber",
                doCreditResponse.hostInformation().retrievalReferenceNumber()
            )
            //      map.putString("SigFileName", doCreditResponse.SigFileName);
            map.putString("SignData", doCreditResponse.signatureInformation().signatureData())
            map.putString("Timestamp", doCreditResponse.traceInformation().timeStamp())
            map.putString("Track1Data", doCreditResponse.accountInformation().track1Data())
            map.putString("Track2Data", doCreditResponse.accountInformation().track2Data())
            map.putString("Track3Data", doCreditResponse.accountInformation().track3Data())
            map.putString(
                "TransactionIntegrityClass",
                doCreditResponse.hostInformation().transactionIntegrityClass()
            )
            map.putString(
                "TransactionRemainingAmount",
                doCreditResponse.amountInformation().transactionRemainingAmount()
            )

            //      map.putString("ExtData", doCreditResponse.ExtData);
            promise.resolve(map)
        } else {
            val errorMsg = executionResult.message()
            val errorCode = executionResult.code().toString()
            Log.d("PAYMENT_RESPONSE_DEBUG", errorMsg)
            Log.d("PAYMENT_RESPONSE_DEBUG", errorCode)

            promise.reject(errorCode, errorMsg)
        }
    }

    private fun handleCashExecutionResult(
        executionResult: ExecutionResult<DoCashResponse>,
        promise: Promise
    ) {
        if (executionResult.isSuccessful) {
            val doCashResponse = executionResult.response()
            Log.d("PAYMENT_RESPONSE_DEBUG AuthCode", doCashResponse.responseCode())
            val map = Arguments.createMap()
            map.putString("ApprovedAmount", doCashResponse.amountInformation().approvedAmount())
            map.putString(
                "ApprovedCashBackAmount",
                doCashResponse.amountInformation().approvedCashBackAmount()
            )
            map.putString(
                "ApprovedMerchantFee",
                doCashResponse.amountInformation().approvedMerchantFee()
            )
            map.putString(
                "ApprovedTaxAmount",
                doCashResponse.amountInformation().approvedTaxAmount()
            )
            map.putString(
                "ApprovedTipAmount",
                doCashResponse.amountInformation().approvedTipAmount()
            )
            map.putString("AuthCode", doCashResponse.responseCode())
            map.putString("AuthorizationResponse", doCashResponse.responseMessage())
            //      map.putString("AvsResponse", doCashResponse.avsInformation().avsMessage());
//      map.putString("CurrentAccountNumber", doCashResponse.accountInformation().currentAccountNumber());
//      map.putString("CardType", doCashResponse.accountInformation().cardType());
//      map.putString("CvdApprovalCode", doCashResponse.accountInformation().cvdApprovalCode());
//      map.putString("DebitAccountType", doCashResponse.accountInformation().debitAccountType());
            map.putString("ECRTransID", doCashResponse.hostTraceInformation().ecrTransactionId())
//            map.putString("EDCType", doCashResponse.edcType())
            //      map.putString("ExtraBalance", doCreditResponse.amountInformation().ExtraBalance);
            map.putString(
                "GatewayTransactionID",
                doCashResponse.hostInformation().gatewayTransactionId()
            )
            //      map.putString("GiftCardType", doCashResponse.accountInformation().giftCardType());
            map.putString("ReferenceNumber", doCashResponse.traceInformation().referenceNumber())
            //      map.putString("HostCardType", doCashResponse.accountInformation().hostCardType());
            map.putString("HostResponseCode", doCashResponse.hostInformation().hostResponseCode())
            map.putString(
                "HostResponseMessage",
                doCashResponse.torInformation().hostResponseMessage()
            )
            map.putString(
                "HostDetailMessage",
                doCashResponse.hostInformation().hostDetailedMessage()
            )
            map.putString("HostTimeStamp", doCashResponse.hostTraceInformation().hostTimeStamp())
            map.putString(
                "HostReferenceNumber",
                doCashResponse.hostTraceInformation().hostReferenceNumber()
            )
            map.putString(
                "IssuerResponseCode",
                doCashResponse.hostInformation().issuerResponseCode()
            )
            map.putString("MaskedPAN", doCashResponse.torInformation().maskedPan())
            map.putString("ResponseMessage", doCashResponse.responseMessage())
            //      map.putString("PayloadData", doCashResponse.payloadData());
            map.putString(
                "PaymentAccountReferenceID",
                doCashResponse.hostInformation().paymentAccountReferenceId()
            )
            //      map.putString("RawResponse", doCashResponse.RawResponse);
//      map.putString("RemainingBalance", doCashResponse.RemainingBalance);
            map.putString("RequestedAmount", doCashResponse.torInformation().originalAmount())
            //      map.putString("ResultCode", doCashResponse.ResultCode);
//      map.putString("ResultTxt", doCashResponse.ResultTxt);
            map.putString(
                "RetrievalReferenceNumber",
                doCashResponse.hostInformation().retrievalReferenceNumber()
            )
            //      map.putString("SigFileName", doCashResponse.SigFileName);
            map.putString("SignData", doCashResponse.signatureInformation().signatureData())
            map.putString("Timestamp", doCashResponse.traceInformation().timeStamp())
            //      map.putString("Track1Data", doCashResponse.accountInformation().track1Data());
//      map.putString("Track2Data", doCashResponse.accountInformation().track2Data());
//      map.putString("Track3Data", doCashResponse.accountInformation().track3Data());
            map.putString(
                "TransactionIntegrityClass",
                doCashResponse.hostInformation().transactionIntegrityClass()
            )
            map.putString(
                "TransactionRemainingAmount",
                doCashResponse.amountInformation().transactionRemainingAmount()
            )

            //      map.putString("ExtData", doCashResponse.ExtData);
            promise.resolve(map)
        } else {
            val errorMsg = executionResult.message()
            val errorCode = executionResult.code().toString()
            Log.d("PAYMENT_RESPONSE_DEBUG", errorMsg)
            Log.d("PAYMENT_RESPONSE_DEBUG", errorCode)

            promise.reject(errorCode, errorMsg)
        }
    }

    companion object {
        const val NAME: String = "PaxPoslink2"
    }
}
